paddle/fluid/framework/unused_var_check.cc

```diff
@@ -0,0 +1,125 @@
+
+DEFINE_bool(enable_unused_var_check, false,
+            "Checking whether operator contains unused inputs, "
+            "especially for grad operator. It should be in unittest.");
+
+const std::unordered_set<std::string> op_has_unsed_vars_white_list = {
+    "auc",
+    "batch_norm",
+    "batch_norm_grad",
+    "sync_batch_norm_grad",
+    "center_loss_grad",
+    "crop",
+    "cvm",
+    "cos_sim_grad",
+    "dgc_momentum",
+    "fake_quantize_range_abs_max",
+    "fill_zeros_like",
+    "fusion_seqpool_cvm_concat",
+    "reshape2_grad_grad",
+    "reshape2_grad",
+    "gru_grad",
+    "hierarchical_sigmoid_grad",
+    "nce_grad",
+    "roi_perspective_transform_grad",
+    "sequence_conv_grad",
+    "gru_unit_grad",
+    "affine_grid_grad",
+    "fill_any_like",
+    "precision_recall",
+    "unsqueeze_grad",
+    "kldiv_loss_grad",
+    "cvm_grad",
+    "stack_grad",
+    "warpctc_grad",
+    "sync_batch_norm",
+    "match_matrix_tensor_grad",
+    "ngraph_engine"};
+
+namespace paddle {
+namespace framework {
+
+std::unordered_set<std::string> *GetThreadLocalUsedVarNameSet() {
+  thread_local std::unordered_set<std::string> used_var_name_set;
+  return &used_var_name_set;
+}
+
+void LogVarUsageIfUnusedVarCheckEnabled(const std::string &name) {
+  VLOG(6) << "Variable used:" << name;
+  GetThreadLocalUsedVarNameSet()->insert(name);
+}
+
+void CheckUnusedVar(const OperatorBase &op, const Scope &scope) {
+  // skip op in white list and it should be fixed in the future.
+  if (op_has_unsed_vars_white_list.count(op.Type()) != 0) {
+    return;
+  }
+  auto *used_set = GetThreadLocalUsedVarNameSet();
+  std::vector<std::string> unsed_input_var_names;
+  auto &inferer = op.Info().NoNeedBufferVarsInferer();
+  std::unordered_set<std::string> no_need_buffer_ins = {};
+  if (inferer) {
+    no_need_buffer_ins = inferer(op.Inputs(), op.Outputs(), op.Attrs());
+  }
+
+  for (auto &pair : op.Inputs()) {
+    // skip no need buffer vars declared
+    if (no_need_buffer_ins.count(pair.first) != 0) {
+      VLOG(6) << op.Type() << " " << pair.first;
+      continue;
+    }
+    if (used_set->count(pair.first) == 0) {
+      for (auto &in_var_name : pair.second) {
+        auto *in_var = scope.FindVar(in_var_name);
+        if (in_var != nullptr && in_var->IsInitialized()) {
+          auto *tensor = &in_var->Get<LoDTensor>();
+          if (tensor != nullptr && tensor->IsInitialized()) {
+            unsed_input_var_names.emplace_back(pair.first);
+            break;
+          }
+        }
+      }
+    }
+  }
+  if (!unsed_input_var_names.empty()) {
+    std::string err_msg = "Operator " + op.Type() + " has input(s) not uesed: ";
+    for (auto &in_var_name : unsed_input_var_names) {
+      err_msg += in_var_name;
+      err_msg += ", ";
+    }
+    err_msg +=
+        "please make sure it(them) is(are) needed. If not, remove it(them) "
+        "from inputs; if yes, register NoNeedBufferVars or add the operator to "
```

Comment:
register NoNeedBufferVars or remove them in the GradOpMaker or add the operator to white list in unused_var_check.cc?

Reply:
Discussed offline. I use `if` to discussed failure of unused_var_check on two case, 

1. If the `input var` is not need in computation, the only fix suggestion is "remove them in the inputs."
2. If the `input var` is really need in computation but failed on check (due to some reasons, like different kernels, conditional usage, or only needing dims), the fix suggestion is "register NoNeedBufferVars or add the operator to white list in unused_var_check.cc".